### PR TITLE
[PFS] Session support

### PR DESCRIPTION
**Pull request by roopakv**
Check / use sessions & rollout percentage for remote feature switches.

Fixes: Part of https://app.asana.com/0/1141250355664307/1145091162039545

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [x] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] I have created or modified unit tests to sufficiently cover my changes.

### CODE REVIEW

File: Helper/FeatureSwitch/Decider.php [hunk 1]
```diff
@@ -29,19 +33,131 @@ class Decider extends AbstractHelper {
...
     private $_manager;
 
+    /** @var State */
+    private $_state;
+
+    /** @var CoreSession */
+    private $_session;
+
+    /**
+     * @var fsRepo
```
Comment (matt-thomason) [inline Helper/FeatureSwitch/Decider.php]: This should be `@var FeatureSwitchRepository`

File: Helper/FeatureSwitch/Decider.php [hunk 2]
```diff
@@ -29,19 +33,131 @@ class Decider extends AbstractHelper {
...
+        $this->_session = $coreSession;
+        $this->_state = $state;
         $this->_manager = $manager;
+        $this->_fsRepo = $fsRepo;
+        $this->_fsFactory = $fsFactory;
+    }
+
+    /**
+     * This method returns if a feature switch is enabled for a user.
+     * The way this is conputed is as follows:
```
Comment (matt-thomason) [inline Helper/FeatureSwitch/Decider.php]: computed

File: Helper/FeatureSwitch/Decider.php [hunk 3]
```diff
@@ -29,19 +33,131 @@ class Decider extends AbstractHelper {
...
+        $switch = $this->_fsFactory->create();
+        $switch->setName($switchDef[Definitions::NAME_KEY]);
+        $switch->setValue($switchDef[Definitions::VAL_KEY]);
+        $switch->setDefaultValue($switchDef[Definitions::DEFAULT_VAL_KEY]);
+        $switch->setRolloutPercentage($switchDef[Definitions::ROLLOUT_KEY]);
+        return $switch;
     }
 
+    /**
+     * The returns if the switch is enabled.
```
Comment (matt-thomason) [inline Helper/FeatureSwitch/Decider.php]: This!

File: Helper/FeatureSwitch/Decider.php [hunk 4]
```diff
@@ -29,19 +33,131 @@ class Decider extends AbstractHelper {
...
+        $switch = null;
+        try {
+            $switch = $this->_fsRepo->getByName($switchName);
+        } catch (NoSuchEntityException $e) {
+            // Switch is not in DB. Fall back to defaults.
+            $switch = $this->_switchFromConst($defaultDef);
+        }
+
+        if (!$switch) {
+            error_log("none");
```
Comment (matt-thomason) [inline Helper/FeatureSwitch/Decider.php]: This log doesn't have useful output. Also words in the following comment.

**Comment by roopakv** [comment]
@matt-thomason thank you for the review. it was 3am and clearly my comments were gibberish. cleaned up

**Comment by matt-thomason** [comment]
3AM PRs are the funnest kind to read 👀

**Pull request merged**